---
title: "DNA Methylation Deconvolution Protocol - Ewing RRBS cohort"
author: Michael Scherer, Petr Nazarov, Reka Toth, Shashwat Sahay, Tony Kamoa, Valentin
  Maurer, Nikita Vedenev, Christoph Plass, Thomas Lengauer, Joern Walter, and Pavlo Lutsik
date: "May 03, 2020"
output:
  html_document: default
  pdf_document: default
bibliography: bibliography.bib
---

# Installation and Data Retrieval
The data were generated in a [recent study](https://www.nature.com/articles/nm.4273) which are also available from GEO, accession number [GSE88826](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE88826).

# Data Import

[*RnBeads*](https://rnbeads.org) also support bisulfite sequencing data, provided that the data has been processed to generate single-CpG methylation calls in BED or similar format. However, since *RnBeads* creates a large internal data structure, more leightweight packages can also be used. The input to the remaining steps should be a DNA methylation data matrix.

```{r import, eval=FALSE}
suppressPackageStartupMessages(library(RnBeads))
rnb.options(
  assembly = "hg38",
  identifiers.column = "sample_id",
  import = TRUE,
  import.default.data.type = "bed.dir",
  import.bed.style = "EPP",
  import.table.separator = ",",
  import.sex.prediction = TRUE,
  qc = TRUE,
  preprocessing = FALSE,
  exploratory = FALSE,
  inference = FALSE,
  differential = FALSE,
  export.to.bed = FALSE,
  export.to.trackhub = NULL,
  export.to.csv = FALSE
)
sample.anno <- "annotation/sample_annotation.csv"
bed.folder <- "bed/"

dir.report <- paste0("report",Sys.Date(),"/")
temp.dir <- tempdir()
options(fftempdir=temp.dir)
data.source <- c(bed.folder,sample.anno,"filename")
rnb.set <- rnb.run.analysis(dir.reports = dir.report, data.source=data.source)
# only use Ewing tissue samples
rem.samples <- !grepl("EwS_T",samples(rnb.set))
# remove FFPE samples from the analysis
rem.samples <- rem.samples | samples(rnb.set) %in% c("EwS_T122","EwS_T123","EwS_T124",
"EwS_T125","EwS_T126","EwS_T127","EwS_T129","EwS_T130","EwS_T131","EwS_T132","EwS_T133")
rnb.set <- remove.samples(rnb.set,rem.samples)
```

# Preprocessing and Filtering
[*DecompPipeline*](https://github.com/CompEpigen/DecompPipeline) also support preprocessing of bisulfite sequencing data. Here, we remove all sites that have a read coverage lower than 5 in any of the samples. Additionally, we remove high and low coverage outliers, sites with missing values, annotated SNPs, and sites on the sex chromosomes.

```{r preprocessing, eval=FALSE}
suppressPackageStartupMessages(library(DecompPipeline))
data.prep <- prepare.data.BS(rnb.set = rnb.set,
                          analysis.name = "Ewing_pipeline",
                          filter.coverage = TRUE,
                          min.coverage = 5,
                          min.covg.quant = 0.001,
                          max.covg.quant = 0.999,
                          filter.na = TRUE,
                          filter.snp = TRUE,
                          filter.sex.chromosomes = TRUE,
                          execute.lump = TRUE)
```

# Feature Selection
We select the 5,000 most variably methylated CpGs across the samples for downstream analysis.

```{r feature_selection, eval=FALSE}
cg_subset <- prepare.CG.subsets(rnb.set=data.prep$rnb.set.filtered,
                                marker.selection = "var",
                                n.markers = 5000)
```

# Deconvolution
Deconvolution is applied analogously to the [general protocol](protocol.html). [*RefFreeCellMix*](https://cran.r-project.org/web/packages/RefFreeEWAS/index.html) and [*EDec*](https://github.com/BRL-BCM/EDec) can be used as alternative deconvolution methods.

```{r deconvolution, eval=FALSE}
md.res <- start.medecom.analysis(
  rnb.set=data.prep$rnb.set.filtered,
  cg_groups = cg_subset,
  Ks=2:15,
  lambda.grid = c(0,10^-(2:5)),
  factorviz.outputs = TRUE,
  analysis.name = "Ewing_pipeline",
  cores = 15
)
```

# Interpretation
The returned *MeDeComSet* and associated deconvolution results can directly be imported into [*FactorViz*](https://github.com/CompEpigen/FactorViz) for visualization. Additional interpretation steps can be performed according to the [general protocol](protocol.html).
```{r visualization, eval=FALSE}
suppressPackageStartupMessages(library(FactorViz))
startFactorViz(file.path(getwd(),"TCGA_Ewing_pipeline","FactorViz_outputs"))
```
